crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+                        // position had a merged cell, had a stroke
+                        // of 'None', or because this is the first track.
+                        // Create a new segment to draw. We start spanning this
+                        // track.
+                        current_segment = Some((stroke, offset, size));
+                        // Nothing to yield for this track. The new segment
+                        // might still be extended in the next track.
+                        None
+                    }
+                } else {
+                    // We shouldn't draw here (stroke of None), so we yield the
+                    // current segment, as it was interrupted.
+                    current_segment.take()
+                };
+                offset += size;
+                interrupted_segment
+            } else {
+                // Reached the end of all tracks, so we interrupt and finish
+                // the current segment.
+                current_segment.take()
+            }
+        },
+    )
+}
+
+/// Returns the correct stroke with which to draw a vline right before column
+/// 'x' when going through row 'y', given the stroke of the user-specified line
+/// at this position, if any.
+/// If the vline would go through a colspan, returns None (shouldn't be drawn).
+/// If the one (when at the border) or two (otherwise) cells to the left and
+/// right of the vline have right and left stroke overrides, respectively,
+/// then the cells' stroke overrides are folded together with the vline's
+/// stroke (with priority to the vline's stroke, followed by the right cell's
+/// stroke, and, finally, the left cell's) and returned. If, however, the cells
+/// around the vline at this row do not have any stroke overrides, then the
+/// vline's own stroke is directly returned.
+pub(super) fn vline_stroke_at_row(
+    grid: &CellGrid,
+    x: usize,
+    y: usize,
+    stroke: Option<Arc<Stroke<Abs>>>,
+) -> Option<Arc<Stroke<Abs>>> {
+    if x != 0 && x != grid.cols.len() {
+        // When the vline isn't at the border, we need to check if a colspan would
+        // be present between columns 'x' and 'x-1' at row 'y', and thus overlap
+        // with the line.
+        // To do so, we analyze the cell right after this vline. If it is merged
+        // with a cell before this line (parent_x < x) which is at this row or
+        // above it (parent_y <= y), this means it would overlap with the vline,
+        // so the vline must not be drawn at this row.
+        let first_adjacent_cell = if grid.has_gutter {
+            // Skip the gutters, if x or y represent gutter tracks.
+            // We would then analyze the cell one column after (if at a gutter
+            // column), and/or one row below (if at a gutter row), in order to
+            // check if it would be merged with a cell before the vline.
+            (x + x % 2, y + y % 2)
```

Comment:
We do have a lot of those `if has_gutter { .. } else { .. }` to compute indices. I wonder whether it might make sense to have some helper struct with a bunch of semantic methods to retrieve the coordinates you need in any place.

Reply:
I was thinking of adding some methods to CellGrid related to this, but I haven't thought of a proper design for that right now. I think it's wiser to delay this refactor to a future PR (perhaps rowspans).